compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -5927,10 +5837,149 @@ public void visit(BLangGroupExpr groupExpr) {
     @Override
     public void visit(BLangRecordLiteral recordLiteral) {
         List<RecordLiteralNode.RecordField> fields =  recordLiteral.fields;
+        generateFieldsForUserUnspecifiedRecordFields(recordLiteral, fields);
         fields.sort((v1, v2) -> Boolean.compare(isComputedKey(v1), isComputedKey(v2)));
         result = rewriteExpr(rewriteMappingConstructor(recordLiteral));
     }
 
+    private SymbolEnv findEnclosingInvokableEnv(SymbolEnv env, BLangInvokableNode encInvokable) {
+        if (env.enclEnv.node != null && env.enclEnv.node.getKind() == NodeKind.ARROW_EXPR) {
+            return env.enclEnv;
+        }
+
+        if (env.enclEnv.node != null && (env.enclEnv.node.getKind() == NodeKind.ON_FAIL)) {
+            return env.enclEnv;
+        }
+
+        if (env.enclInvokable != null && env.enclInvokable == encInvokable) {
+            return findEnclosingInvokableEnv(env.enclEnv, encInvokable);
```

Comment:
Why do we need these special cases?

Reply:
Why specifically arrow and on-fail checks?

Reply:
Doesn't the closure desugar already handle these?

Reply:
No, When comes to the closure desugar phase, need to identify and set them as closure variables. (Defining a variable as a closure variable is not done during the closure desugar phase.)

Since we are generating the record fields during the desugar phase, we need to indicate whether the function pointers associated with those invocations are closure variables or not.